### PR TITLE
Search: Added search function by first name and last name

### PR DESCRIPTION
**Pull request by Filip-B-32**
search implemented

### CODE REVIEW

File: collaborative_practice.js [hunk 1]
```diff
@@ -32,6 +42,10 @@ const runProgram = async () => {
                 case "1":
                     listUsers(users);
                     break;
+                case "5":
+                    console.log("Person you are looking for: ")
+                    searchUsers(await prompt("First Name: "), await prompt("Last Name: "), users)
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: Please move the console log, and the two prompts inside searchUsers function

File: collaborative_practice.js [hunk 2]
```diff
@@ -19,6 +20,15 @@ const listUsers = (users) => {
     console.log(output);
 };
 
+const searchUsers = (firstName, lastName, users) => {
+    const output = users
+        .map(user => (user.firstName === firstName && user.lastName === lastName)
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: We need to search based on a single input and check if that input is a substring of firstName or lastName

File: collaborative_practice.js [hunk 3]
```diff
@@ -19,6 +20,15 @@ const listUsers = (users) => {
     console.log(output);
 };
 
+const searchUsers = (firstName, lastName, users) => {
+    const output = users
+        .map(user => (user.firstName === firstName && user.lastName === lastName) 
+        ? `ID: ${user.id}, First name: ${user.firstName}, Last name: ${user.lastName}`
+        : "User not found")
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: We might find more than one user based on the search string

File: collaborative_practice.js (hunk 1 above)
Comment (gheorghe-sava) [reply collaborative_practice.js]: The switch only calls functions for each operation and nothing more

File: collaborative_practice.js (hunk 3 above)
Comment (Filip-B-32) [reply collaborative_practice.js]: Do we want to show the first person with that first name or last name, or do we need to show all of the people with that certain first name or last name?

File: collaborative_practice.js [hunk 4]
```diff
@@ -8,7 +8,8 @@ const printMenu = () => {
     console.log(`
     Select an option:
         1. List all users
-        5. Search user in list of users
+        2. Create new user
+		5. Search user in list of users
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: Bad indentation

File: collaborative_practice.js (hunk 3 above)
Comment (gheorghe-sava) [reply collaborative_practice.js]: We want to show every user with that substring in their first or last name.
Minimum length of the substring must be 2 characters

File: collaborative_practice.js [hunk 5]
```diff
@@ -46,21 +60,20 @@ const runProgram = async () => {
...
                     }
-                    
+
                     users.push(newUser)
                     console.log("User createad succesfully!");
                     break;
                 case "5":
-                    console.log("Person you are looking for: ")
-                    searchUsers(await prompt("First Name: "), await prompt("Last Name: "), users)
+                    searchUsers(await prompt("Please enter a first name or a last name: "), users)
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: Please read the input inside the function

File: collaborative_practice.js [hunk 6]
```diff
@@ -22,18 +22,32 @@ const listUsers = (users) => {
...
+const searchUsers = (name, users) => {
+    if (name.length >= 2) {
+        let exists = false;
+        users
+            .forEach(user => {
+                if (user.firstName.toLowerCase() === name.toLowerCase() || user.lastName.toLowerCase() === name.toLowerCase()) {
+                    console.log(`ID: ${user.id}, First name: ${user.firstName}, Last name: ${user.lastName}\n`)
+                    exists = true;
+                }
+            })
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: Maybe we can reuse the listUsers function, what remains is to filter out the users which do not meet the criteria and pass the resulted list to listUsers

File: collaborative_practice.js [hunk 7]
```diff
@@ -22,18 +22,32 @@ const listUsers = (users) => {
...
+            
+        if (!exists) {
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: We will not need exists boolean if we have a resulted list which has or has not elements in it

File: collaborative_practice.js [hunk 8]
```diff
@@ -22,18 +22,32 @@ const listUsers = (users) => {
...
-        ? `ID: ${user.id}, First name: ${user.firstName}, Last name: ${user.lastName}`
-        : "User not found")
-    
-    console.log(output)
+const searchUsers = (name, users) => {
+    if (name.length >= 2) {
+        let exists = false;
+        users
+            .forEach(user => {
+                if (user.firstName.toLowerCase() === name.toLowerCase() || user.lastName.toLowerCase() === name.toLowerCase()) {
```
Comment (gheorghe-sava) [inline collaborative_practice.js]: We have to check if the name is contained in the first or last name
Eg: "an" matches "Dan" and "Daniels"

File: collaborative_practice.js (hunk 6 above)
Comment (Filip-B-32) [reply collaborative_practice.js]: From the listUsers function we get a string and we could't use the filter method. In my opinion is for the best to use an array. Can i use the users array?

File: collaborative_practice.js (hunk 8 above)
Comment (Filip-B-32) [reply collaborative_practice.js]: i checked, it doesn't show "Dan" or "Daniel" if input is "an"

File: collaborative_practice.js (hunk 6 above)
Comment (gheorghe-sava) [reply collaborative_practice.js]: Inside searchUsers we compute a list of users that meet the criteria of having the search string inside the fist or last name
We call listUsers inside searchUsers and we pass the computed list of users that meet the criteria to listUsers, then listUsers just outputs the filtered users
Indeed, the log message from listUsers is not that consistent with the search but we can ignore it for now

File: collaborative_practice.js (hunk 8 above)
Comment (gheorghe-sava) [reply collaborative_practice.js]: Then we have to make it work so users with ID 1 and 2 will be shown when the user searches for "an"

File: collaborative_practice.js [hunk 9]
```diff
@@ -22,11 +22,35 @@ const listUsers = (users) => {
...
+            .filter(user => user.firstName.toLowerCase() === name.toLowerCase() 
+                        || user.lastName.toLowerCase() === name.toLowerCase()
+                        || user.firstName.toLowerCase().includes(name)
+                        || user.lastName.toLowerCase().includes(name))
+
+        if (output.length === 0) {
+            console.log("The name does not exist!")
+        }
+        else {
+            console.log(output)
```
Comment (gheorghe-sava) [inline collaborative_practice.js]:
```suggestion
            listUsers(output)
```

**Review by gheorghe-sava — APPROVED** [review]
Looks good!

**Review by HellHound97 — APPROVED** [review]

**Pull request merged**